inc/functions.php

```diff
@@ -549,6 +549,65 @@ function my_date($format, $stamp=0, $offset="", $ty=1, $adodb=false)
 	return $date;
 }
 
+/**
+ * Get a mail handler instance, a MyBB's built-in SMTP / PHP mail hander or one created by a plugin.
+ * @param bool $use_buitlin Whether to use MyBB's built-in mail handler.
+ *
+ * @return object A MyBB built-in mail handler or one created by plugin(s).
+ */
+function &get_my_mailhandler($use_buitlin = false)
+{
+	global $mybb, $plugins;
+	static $my_mailhandler;
+	static $my_mailhandler_builtin;
+
+	if($use_buitlin)
+	{
+		// If our built-in mail handler doesn't exist, create it.
+		if(!is_object($my_mailhandler_builtin))
+		{
+			require_once MYBB_ROOT . "inc/class_mailhandler.php";
+
+			// Using SMTP.
+			if(isset($mybb->settings['mail_handler']) && $mybb->settings['mail_handler'] == 'smtp')
+			{
+				require_once MYBB_ROOT . "inc/mailhandlers/smtp.php";
+				$my_mailhandler_builtin = new SmtpMail();
+			}
+			// Using PHP mail().
+			else
+			{
+				require_once MYBB_ROOT . "inc/mailhandlers/php.php";
+				$my_mailhandler_builtin = new PhpMail();
+				if(!empty($mybb->settings['mail_parameters']))
+				{
+					$my_mailhandler_builtin->additional_parameters = $mybb->settings['mail_parameters'];
+				}
+			}
+		}
+
+		return $my_mailhandler_builtin;
+	}
+
+	// If our mail handler doesn't exist, create it.
+	if(!is_object($my_mailhandler))
+	{
+		require_once MYBB_ROOT . "inc/class_mailhandler.php";
+
+		$plugins->run_hooks('my_mailhandler_init', $my_mailhandler);
+
+		// If no plugin creates the mail handler, resort to use the built-in one.
+		if(!is_object($my_mailhandler) || !($my_mailhandler instanceof MailHandler))
+		{
+			$my_mailhandler = &get_my_mailhandler(true);
+		}
+
+		$plugins->run_hooks('my_mailhandler_after_init', $my_mailhandler);
```

Comment:
I'm now being picky, but we usually use the `_start` and `_end` suffixes for hooks that need it. Shouldn't be something to stop this from being merged of course.

Reply:
OK, I'll rename them accordingly shortly. However, maybe just like you talked about using `get_my_mailhandler(true)`, here the latter hook may be there right after the former hook initialized a custom handler with the in-between `if-block` unsatisfied. I just don't know what exactly should be done here right now.

Reply:
Well, I think it's here to discuss about this comment:

> I think this assumes get_my_mailhandler(true) will never be used by third parties, which might not be the case.

Yes, that's right, maybe.

If no plugin hooks to create a custom handler before calling my_mail, then nothing can be done to create a custom handler. So I'm thinking about returning the builtin one when `use_builtin` is `false other than assigning it to `$my_mailhandler`, which means that just leave it as empty.

Reply:
I think you might be right about the hook mentioned above being unnecessary at this moment, I have  resolved the conversation. Developers could use `get_my_mailhandler()` and get the built in handler overwritten that way.

> So I'm thinking about returning the builtin one when use_builtin is false other than assigning it to $my_mailhandler`, which means that just leave it as empty.

Having resolved my opinion on the above issue, I think this hook should stay as is (to assign it to and return `$my_mailhandler`).

If you think any of my points is still valid for further consideration feel free to redo the resolves.